packages/vizbuilder/src/index.js

```diff
@@ -0,0 +1,112 @@
+import React from "react";
+import PropTypes from "prop-types";
+import classnames from "classnames";
+import queryString from "query-string";
+
+import {fetchCubes} from "./actions/fetch";
+import {loadControl, setStatePromise} from "./actions/loadstate";
+import AreaChart from "./components/AreaChart";
+import AreaLoading from "./components/AreaLoading";
+import AreaSidebar from "./components/AreaSidebar";
+import {UIToaster} from "./components/UIToaster";
+import * as api from "./helpers/api";
+import initialState from "./state";
+
+import "@blueprintjs/labs/dist/blueprint-labs.css";
+import "./index.css";
+
+class Vizbuilder extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    api.resetClient(props.src);
+    this.state = initialState();
+
+    this.loadControl = loadControl.bind(this);
+    this.firstLoad = this.firstLoad.bind(this);
+    this.stateUpdate = this.stateUpdate.bind(this);
+  }
+
+  getChildContext() {
+    return {
+      loadControl: this.loadControl,
+      stateUpdate: this.stateUpdate
+    };
+  }
+
+  componentDidMount() {
+    const locationQuery = queryString.parse(location.search);
+    this.firstLoad(locationQuery);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const {src, datasetDidChange} = this.props;
+    const {dataset, load} = this.state;
+    const {error, severity} = load;
+
+    if (src && prevProps.src !== src) {
+      api.resetClient(src);
+      this.setState(initialState(), this.firstLoad);
+    }
+
+    if (datasetDidChange && dataset !== prevState.dataset) {
+      datasetDidChange(dataset);
+    }
+
+    if (error && prevState.load.error !== error) {
+      console.warn(error.stack);
+      UIToaster.show({intent: severity, message: error.message});
+    }
+  }
+
+  render() {
+    const {dataset, load, members, options, query, queryOptions} = this.state;
+    return (
+      <div className={classnames("vizbuilder", {loading: load.inProgress})}>
+        <AreaLoading {...load} />
+        <AreaSidebar options={options} query={query} queryOptions={queryOptions} />
+        <AreaChart dataset={dataset} members={members} query={query} />
+      </div>
+    );
+  }
+
+  stateUpdate(newState) {
```

Comment:
Nope, because this code is combining elements exactly one level deep in the state.
If `newState` is `{query: {conditions: newConditions}}`, this code merges the elements inside `query`.